### ISSUE TITLE
Use nearest as NearestNeighbors 

### ISSUE THREAD

**Issue by pvnieo** [rusty1s/pytorch_cluster#82]
Hi,

I want to use `torch_cluster` as nearest neighbor algorithm, like scikit-lean's [NearestNeighbors](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.NearestNeighbors.html#sklearn.neighbors.NearestNeighbors), or scipy's [KDTree](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.KDTree.query.html), and I want to know if this possible, and if so, has an accuracy assessment been done?

In fact, when I test `torch_cluster.nearest`, it gives me an accuracy lower than the algorithm above, and I want to know if this normal (due to the algorithm used) or I was doing something wrong!

Thank you in advance!

**Comment by rusty1s** [comment]
Can you show me a small example to reproduce this? Note that `torch_cluster.nearest` only returns the single nearest point, and for more neighbors, you may want to use `torch_cluster.knn`.

**Comment by pvnieo** [comment]
Hi,

Thank you for your response. Yes, In fact, I'm only interested in the closest neighbor.

Here's a reproduction example:
```
from scipy.spatial import cKDTree
from sklearn.neighbors import NearestNeighbors
from torch_cluster import nearest

A, B = torch.load("example.pt")

neigh = NearestNeighbors(n_jobs=-1)
neigh.fit(A)
nn_sklearn = neigh.kneighbors(B, 1, return_distance=False).flatten()

tree = cKDTree(A)
nn_scipy = tree.query(B)[1]
(nn_scipy == nn_sklearn).sum() / nn_sklearn.shape[0]  # 1.0

nn_torch = nearest(torch.Tensor(B), torch.Tensor(A)).numpy()
(nn_torch == nn_sklearn).sum() / nn_sklearn.shape[0], (nn_torch == nn_scipy).sum() / nn_scipy.shape[0]  # (0.9854116706634692, 0.9854116706634692)
```

**Comment by pvnieo** [comment]
Also, you mentioned in your response that there is a function `torch_cluster.knn`, but it isn't documented in the github page. Is this because it still in experimental phase?

**Comment by pvnieo** [comment]
Hi @rusty1s did the reproduction example work for you (the sharing link is ok?)?

**Comment by rusty1s** [comment]
Hey, sorry for late reply and the reminder! I don't have rights to access the google drive, but the following code works without any issues for me:
```python
from sklearn.neighbors import NearestNeighbors
import torch
from torch_cluster import nearest

torch.manual_seed(12345)

A = torch.randn(100, 3)
B = torch.randn(100, 3)

neigh = NearestNeighbors(n_jobs=-1)
neigh.fit(A)
nn_sklearn = neigh.kneighbors(B, 1, return_distance=False).flatten()

nn_torch = nearest(B, A)
assert nn_sklearn.tolist() == nn_torch.tolist()
```

**Comment by pvnieo** [comment]
@rusty1s Hi, sorry for the link problem, I just gave you access to it!

**Comment by rusty1s** [comment]
Ok, I had a look and this is fixed in master now. It was caused by numerical instabilities since we re-normalize points to `[0,1]` to handle batches of points.

**Comment by pvnieo** [comment]
Hi @rusty1s 
Thanks for the quick reply, I will test it and let you now!

**Comment by pvnieo** [comment]
Hi @rusty1s 

How can I install the version in master?

**Comment by rusty1s** [comment]
The easiest is just to apply [the fix](https://github.com/rusty1s/pytorch_cluster/commit/c5045e2fd3dfdcf5c3e8c626164c924efba17698) locally. No need to re-install all kernels.

**Comment by pvnieo** [comment]
Hi @rusty1s I applied the fix locally (changed the code directly in the site-packages), but nothing changed!
Is this because the kernels are already compiled (`_nearest.so`)?

**Comment by rusty1s** [comment]
Can you directly change the content of the file with the [this one](https://github.com/rusty1s/pytorch_cluster/blob/master/torch_cluster/nearest.py). I made some further adjustments.

**Comment by pvnieo** [comment]
Hi @rusty1s 
Yes, this change worked for me, now the precision changed from **0.9854116706634692** to **0.999400479616307** !

it's still not 1 because of the same numerical instabilities?

**Comment by pvnieo** [comment]
Hi @rusty1s
Thank you for including the fix in the new release!

I tested with the new version, and I still get **0.9992006394884093**, and I want to be sure that this is because of the numerical instabilities?
